code/modules/integrated_electronics/subtypes/reagents.dm

```diff
@@ -33,52 +45,141 @@
 		for(var/i = 1 to 8)
 			smoke_system.start()
 		reagents.clear_reagents()
+	activate_pin(2)
 
 /obj/item/integrated_circuit/reagent/injector
 	name = "integrated hypo-injector"
 	desc = "This scary looking thing is able to pump liquids into whatever it's pointed at."
 	icon_state = "injector"
 	extended_desc = "This autoinjector can push reagents into another container or someone else outside of the machine.  The target \
-	must be adjacent to the machine, and if it is a person, they cannot be wearing thick clothing."
+	must be adjacent to the machine, and if it is a person, they cannot be wearing thick clothing.Negative given amount makes injector suck out reagents."
 	flags = OPENCONTAINER
 	complexity = 20
 	cooldown_per_use = 6 SECONDS
 	inputs = list("target" = IC_PINTYPE_REF, "injection amount" = IC_PINTYPE_NUMBER)
 	inputs_default = list("2" = 5)
-	outputs = list()
-	activators = list("inject" = IC_PINTYPE_PULSE_IN)
+	outputs = list("volume used" = IC_PINTYPE_NUMBER,"self reference" = IC_PINTYPE_REF)
+	activators = list("inject" = IC_PINTYPE_PULSE_IN, "on injected" = IC_PINTYPE_PULSE_OUT, "on fail" = IC_PINTYPE_PULSE_OUT)
 	spawn_flags = IC_SPAWN_DEFAULT|IC_SPAWN_RESEARCH
 	volume = 30
 	power_draw_per_use = 15
+	var/direc = 1
+	var/transfer_amount = 10
+
+/obj/item/integrated_circuit/reagent/injector/interact(mob/user)
+	set_pin_data(IC_OUTPUT, 2, weakref(src))
+	push_data()
+	..()
+
+
+/obj/item/integrated_circuit/reagent/injector/on_reagent_change()
+	set_pin_data(IC_OUTPUT, 1, reagents.total_volume)
+	push_data()
+
+/obj/item/integrated_circuit/reagent/injector/on_data_written()
+	var/new_amount = get_pin_data(IC_INPUT, 2)
+	if(new_amount < 0)
+		new_amount = -new_amount
+		direc = 0
+	else
+		direc = 1
+	if(isnum(new_amount))
+		new_amount = Clamp(new_amount, 0, volume)
+		transfer_amount = new_amount
 
-/obj/item/integrated_circuit/reagent/injector/proc/inject_amount()
-	var/amount = get_pin_data(IC_INPUT, 2)
-	if(isnum(amount))
-		return Clamp(amount, 0, 30)
 
 /obj/item/integrated_circuit/reagent/injector/do_work()
 	set waitfor = 0 // Don't sleep in a proc that is called by a processor without this set, otherwise it'll delay the entire thing
-
 	var/atom/movable/AM = get_pin_data_as_type(IC_INPUT, 1, /atom/movable)
 	if(!istype(AM)) //Invalid input
+		activate_pin(3)
 		return
-	if(!reagents.total_volume) // Empty
-		return
-	if(AM.can_be_injected_by(src))
-		if(isliving(AM))
-			var/mob/living/L = AM
-			var/turf/T = get_turf(AM)
-			T.visible_message("<span class='warning'>[src] is trying to inject [L]!</span>")
-			sleep(3 SECONDS)
-			if(!L.can_be_injected_by(src))
+
+	if(direc == 1)
```

Comment:
Would it be possible to trim this code down, since it looks like it was copypasted twice to enable two-directional injection?

Reply:
I'll see about it tomorrow.